src/components/TodoList.jsx

```diff
@@ -1,66 +1,239 @@
+                }))
+                : task
+            )
+        })
+    }
+    // Set text deoration style of task
+    setTaskTextDecoration(taskId) {
+        const { tasks } = this.state
+        this.setState({
+            ...this.state,
+            tasks: tasks.map(task => task.id === taskId
+                ? (({
+                    ...task,
+                    bIsDone: !task.bIsDone,
+                    textDecoration: !task.bIsDone ? "line-through" : ""
+                }))
+                : task
+            )
+        })
     }
 
-    removeTask(text) {
+    // Set text of task in edit mode 
+    setTaskInEditId(todoId) {
+        this.setState({
+            ...this.state,
+            taskInEditId: todoId
+        })
+    }
+    // Change name of task
+    setNewTaskText(newText) {
+        const { tasks, taskInEditId } = this.state
+        this.setState({
+            ...this.state,
+            tasks: tasks.map(task => task.id === taskInEditId
+                ? ({ ...task, taskName: newText })
+                : task),
+            taskInEditId: -1
+        })
+    }
+    // Update state with sorted tasks
+    sortTasks(event) {
         const { tasks } = this.state;
+        const option = event.target.value
+        this.setState({
+            ...this.state,
+            tasks: tasks.sort(filterTasks(option)),
+            sortingOption: option
+        });
+    }
 
-
+    // Remove task
+    removeTask(taskId) {
+        const { tasks } = this.state;
         const newState = {
             ...this.state,
-            tasks: tasks.filter(elem => elem != text)
-        }
+            tasks: tasks.filter(task => task.id !== taskId)
+        };
         this.setState(newState);
     }
-
+    // Add task
     addTask() {
-        const { tasks } = this.state;
-
+        const { tasks, sortingOption, lastCreatedId, input } = this.state;
+        if (input.trim().length === 0)
+            return
+        // Set properties of the new task
+        const newTask = {
+            taskName: input,
+            date: new Date(),
+            id: lastCreatedId + 1,
+            color: "",
+            textDecoration: "",
+            bIsDone: false,
+            bIsImportant: false
+        }
+        Object.freeze(newTask)
+        // Sort task as it added 
+        const newTasks = tasks.concat(newTask).sort(filterTasks(sortingOption));
         this.setState({
             input: "",
-            tasks: tasks.concat(this.state.input)
+            tasks: newTasks,
+            lastCreatedId: lastCreatedId + 1
         });
     }
 
     onInputChange(event) {
         const text = event.target.value;
-        this.setState({ ...this.state, input: text });
+        this.setState({
+            ...this.state,
+            input: text
+        })
     }
 
+    removeAllTasks() {
+        this.setState(
+            {
+                ...this.state,
+                tasks: []
```

Comment:
This is where Typescript would've told you you made a mistake. You are setting tasks to an empty array, while this should be the same type as tasks were before cleanup - Immutable.js list :)